rust/template/d3log/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+    fn localize(&self, rel: usize, v: DDValue) -> Option<(Node, usize, DDValue)>;
+    fn now(&self) -> u64;
+    fn myself(&self) -> Node;
+    fn error(&self, text: Record, line: Record, filename: Record, functionname: Record);
+    fn record_from_ddvalue(&self, d: DDValue) -> Result<Record, Error>;
+    fn relation_name_from_id(&self, id: usize) -> Result<String, Error>;
+    fn relation_deserializer(&self, id: usize) -> Result<AnyDeserializeSeed, Error>;
+}
+
+pub type Evaluator = Arc<(dyn EvaluatorTrait + Send + Sync)>;
+
+#[derive(Clone)]
+pub struct Instance {
+    pub uuid: Node,
+    pub broadcast: Arc<Broadcast>,
+    pub init_batch: Batch,
+    pub eval_port: Port,
+    pub eval: Evaluator,
+    pub dispatch: Arc<Dispatch>,
+    pub forwarder: Arc<Forwarder>,
+    pub rt: Arc<tokio::runtime::Runtime>,
+}
+
+impl Instance {
+    pub fn new(
+        rt: Arc<Runtime>,
+        new_evaluator: Arc<dyn Fn(Node, Port) -> Result<(Evaluator, Batch), Error> + Send + Sync>,
+        uuid: u128,
+    ) -> Result<Arc<Instance>, Error> {
+        let broadcast = Broadcast::new(uuid);
+        let (eval, init_batch) = new_evaluator(uuid, broadcast.clone())?;
+        Broadcast::register_eval(broadcast.clone(), eval.clone());
+        let dispatch = Arc::new(Dispatch::new(eval.clone()));
+        let forwarder = Forwarder::new(eval.clone(), dispatch.clone(), broadcast.clone());
+
+        let (esend, erecv) = channel(20);
+        let eval_port = Arc::new(EvalPort {
+            rt: rt.clone(),
+            eval: eval.clone(),
+            dispatch: dispatch.clone(),
+            s: esend,
+        });
+
+        let instance = Arc::new(Instance {
+            init_batch,
+            uuid,
+            rt,
+            eval_port: eval_port.clone(),
+            broadcast: broadcast.clone(),
+            forwarder: forwarder.clone(),
+            dispatch: dispatch.clone(),
+            eval: eval.clone(),
+        });
+
+        Instance::eval_loop(instance.clone(), erecv);
+        broadcast.clone().subscribe(eval_port.clone(), u128::MAX);
+        ThreadInstance::new(instance.clone(), new_evaluator)?;
+        eval_port.send(fact!(d3_application::Myself, me => uuid.into_record()));
+        Ok(instance)
+    }
+
+    // doesn't seem work defining a trait just so we can get 'self : Arc<Instance>'
+    fn eval_loop(instance: Arc<Instance>, mut erecv: Receiver<Batch>) {
+        let i2 = instance.clone();
+        i2.clone().rt.spawn(async move {
+            loop {
+                if let Some(b) = erecv.recv().await {
+                    let e = i2.eval.clone();
+                    i2.dispatch.send(b.clone());
+                    let out = async_error!(e.clone(), e.eval(b.clone()));
+                    i2.dispatch.send(out.clone());
+                    i2.forwarder.send(out.clone());
+                } else {
+                    panic!("eval shut down");
+                }
+            }
+        });
+    }
```

Comment:
Personally I'd rather this be an `async` function that the caller then calls `Runtime::spawn()` on, but it's up to you